paymentsheet/src/main/java/com/stripe/android/customersheet/data/CustomerSessionElementsSessionManager.kt

```diff
@@ -0,0 +1,98 @@
+    private var cachedCustomerEphemeralKey: CachedCustomerEphemeralKey? = null
+
+    private var intentConfiguration: CustomerSheet.IntentConfiguration? = null
+
+    override suspend fun fetchCustomerSessionEphemeralKey(): Result<CachedCustomerEphemeralKey> {
+        return withContext(workContext) {
+            cachedCustomerEphemeralKey.takeUnless { cachedCustomerEphemeralKey ->
+                cachedCustomerEphemeralKey == null || cachedCustomerEphemeralKey.shouldRefresh(
+                    timeProvider()
+                )
+            }?.let {
+                Result.success(it)
+            } ?: run {
+                fetchElementsSession().mapCatching {
+                    cachedCustomerEphemeralKey
+                        ?: throw IllegalStateException("Should have been initialized from `elements/session`!")
+                }
+            }
+        }
+    }
+
+    override suspend fun fetchElementsSession(): Result<ElementsSession> {
+        return withContext(workContext) {
+            runCatching {
+                val intentConfiguration = intentConfiguration
+                    ?: customerSessionProvider.intentConfiguration()
+                        .onSuccess { intentConfiguration = it }
+                        .getOrThrow()
+
+                val customerSessionClientSecret = customerSessionProvider
+                    .providesCustomerSessionClientSecret()
+                    .getOrThrow()
+
+                val prefsRepository = prefsRepositoryFactory(customerSessionClientSecret.customerId)
+
+                val savedSelection = prefsRepository.getSavedSelection(
+                    isGooglePayAvailable = false,
+                    isLinkAvailable = false,
+                ) as? SavedSelection.PaymentMethod
+
+                elementsSessionRepository.get(
+                    initializationMode = PaymentSheet.InitializationMode.DeferredIntent(
+                        intentConfiguration = PaymentSheet.IntentConfiguration(
+                            mode = PaymentSheet.IntentConfiguration.Mode.Setup(),
+                            paymentMethodTypes = intentConfiguration.paymentMethodTypes,
+                        )
+                    ),
+                    defaultPaymentMethodId = savedSelection?.id,
+                    customer = PaymentSheet.CustomerConfiguration.createWithCustomerSession(
+                        id = customerSessionClientSecret.customerId,
+                        clientSecret = customerSessionClientSecret.clientSecret,
+                    ),
+                    externalPaymentMethods = listOf(),
+                ).onSuccess { elementsSession ->
+                    elementsSession.customer?.session?.run {
+                        cachedCustomerEphemeralKey = CachedCustomerEphemeralKey(
+                            customerId = customerId,
+                            ephemeralKey = apiKey,
+                            expiresAt = apiKeyExpiry,
+                        )
```

Comment:
Anytime we call `elemenets/session` (on initialization and when fetching payment methods), we should store the latest ephemeral key.

Reply:
After we add validation of the client secret, we should have an unexpected error here if `elementsSession.customer?.session` is null because we expect this to not be null with a valid customer session client secret.

Reply:
How will we validate the client secret? I think we'd want to be very confident that this case will never happen before adding an unexpected error here. I feel like the fact that the model indicates that customer + session can be null makes me feel like this isn't an unexpected error. Though we could definitely have it be an expected error and alert on a low threshold of these errors

Reply:
We would validate the prefix of the key which is [what we are doing with `PaymentSheet`](https://github.com/stripe/stripe-android/blob/master/paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentSheetConfigurationKtx.kt#L46-L65)and the unexpected error that is in there.

Reply:
We expect all `CustomerSession` type keys to return `elementsSession.customer`. Otherwise `CustomerSession` would work improperly.

Reply:
gotcha, makes sense to keep doing this if it's what we do for PaymentSheet! Thanks for the context